File: leetcode-04/solutions/cpp/0098.cpp
class Solution {
 public:
  bool isValidBST(TreeNode* root) { return helper(root, NULL, NULL); }

 private:
  bool helper(TreeNode* root, TreeNode* minNode, TreeNode* maxNode) {
    if (!root) return true;
    if (minNode && root->val <= minNode->val ||
        maxNode && root->val >= maxNode->val)
      return false;

    return helper(root->left, minNode, root) &&
           helper(root->right, root, maxNode);
  }
};

File: leetcode-04/solutions/python3/0098.py
class Solution:
    def isValidBST(self, root: TreeNode) -> bool:
        return self.helper(root, None, None)

    def helper(self, root: TreeNode, minNode: TreeNode, maxNode: TreeNode) -> bool:
        if not root:
            return True
        if minNode and root.val <= minNode.val or maxNode and root.val >= maxNode.val:
            return False

        return self.helper(root.left, minNode, root) and self.helper(root.right, root, maxNode)
